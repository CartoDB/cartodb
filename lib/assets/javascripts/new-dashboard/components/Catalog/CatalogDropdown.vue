<template>
  <div
    class="catalogDropdown"
    @keydown.down.prevent="onKeydownDown"
    @keydown.up.prevent="onKeydownUp"
    :class="{'is-disabled': isDisabled, 'is-open': isOpen}">
    <span class="title is-caption catalogDropdown__label">{{ title }}</span>
    <div class="catalogDropdown__container">
      <div v-if="showInput">
        <input type="text"
          class="text is-caption catalogDropdown__input"
          :class="{ 'has-error': hasError }"
          :placeholder="placeholder"
          v-model="searchFilter"
          @focus="onInputFocus"
          @click="openDropdown"
          @keyup.enter="onKeyEnter"
          :disabled="isDisabled || hasError">
      </div>
      <div v-else
        class="text is-caption catalogDropdown__input"
        :class="{ 'has-error': hasError }"
        @click="openDropdown">
        <CatalogDropdownItem v-if="this.searchFilter" :option="this.searchFilter"/>
        <span v-else class="catalogDropdown__input--placeholder">{{ this.placeholder }}</span>
        <button
          v-if="searchFilter"
          class="catalogDropdown__close"
          @click="reset">
          <img src="../../assets/icons/common/dropdown-close.svg" width="16" height="20" />
        </button>
      </div>
      <p class="catalogDropdown__error text is-small" v-if="hasError">{{ error }}</p>
      <ul ref="catalogDropdownList"
        class="catalogDropdown__list"
        v-click-outside="closeDropdown"
        :class="{
          'is-open ps-container': isOpen,
          'is-height-limited': limitHeight,
          'ps-active-y': showScroll
        }"
        @mouseleave="resetActiveOption">
        <li
          v-for="(option, index) in filteredOptions" :key="option"
          @click="selectOption(option)"
          class="catalogDropdown__option text is-caption"
          :class="{'catalogDropdown__option--active': activeOptionIndex === index + 1}"
          @mouseover="updateActiveOption(index + 1)">
          <CatalogDropdownItem :option="option"/>
        </li>
      </ul>
      <div class="catalogDropdown__extra">
        <slot name="extra"></slot>
      </div>
    </div>
  </div>
</template>

<script>
import CatalogDropdownItem from './CatalogDropdownItem';
import PerfectScrollbar from 'perfect-scrollbar';

export default {
  name: 'CatalogDropdown',
  components: {
    CatalogDropdownItem
  },
  props: {
    title: String,
    placeholderActive: String,
    placeholderInactive: String,
    options: {
      type: Array,
      default () {
        return [];
      }
    },
    open: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    limitHeight: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      selected: {},
      isInputFocused: false,
      activeOptionIndex: -1,
      searchFilter: '',
      isOpen: this.open,
      isDisabled: this.disabled,
      error: ''
    };
  },
  computed: {
    placeholder () {
      return !this.isDisabled ? this.placeholderActive : this.placeholderInactive;
    },
    maxItemsScroll () {
      const PAGE_HEIGHT = window.innerHeight;
      const SMALL_HEIGHT = 680;
      const MEDIUM_HEIGHT = 850;
      const HEIGHT_SMALL_MAX_ITEMS = 3;
      const HEIGHT_MEDIUM_MAX_ITEMS = 5;
      const HEIGHT_DEFAULT_MAX_ITEMS = 8;

      if (PAGE_HEIGHT < SMALL_HEIGHT) {
        return HEIGHT_SMALL_MAX_ITEMS;
      }

      if (PAGE_HEIGHT < MEDIUM_HEIGHT) {
        return HEIGHT_MEDIUM_MAX_ITEMS;
      }

      return HEIGHT_DEFAULT_MAX_ITEMS;
    },
    filteredOptions () {
      const filtered = [];
      const regOption = new RegExp(this.searchFilter, 'ig');
      for (const option of this.options) {
        if (this.searchFilter.length < 1 || option.match(regOption)) {
          filtered.push(option);
        }
      }
      return filtered.length > 0 ? filtered : this.options;
    },
    searchEnabled () {
      return this.$refs.catalogDropdownList
        ? this.$refs.catalogDropdownList.scrollTop === 0
        : false;
    },
    hasError () {
      return this.error.length > 0;
    },
    showScroll () {
      return this.filteredOptions.length > this.maxItemsScroll;
    },
    showInput () {
      return this.isOpen && Object.keys(this.selected).length === 0;
    }
  },
  mounted () {
    this.$refs.catalogDropdownList.scrollTop = 0;
    PerfectScrollbar.initialize(this.$refs.catalogDropdownList, {
      wheelSpeed: 1,
      wheelPropagation: false,
      swipePropagation: true,
      stopPropagationOnClick: false,
      minScrollbarLength: 20,
      useBothWheelAxes: true
    });
  },
  beforeDestroy () {
    PerfectScrollbar.destroy(this.$refs.catalogDropdownList);
  },
  methods: {
    openDropdown (event) {
      if (event) {
        event.stopPropagation();
      }

      if (!this.isOpen && !this.isDisabled) {
        this.isOpen = true;
        this.$refs.catalogDropdownList.scrollTop = 0;
        PerfectScrollbar.update(this.$refs.catalogDropdownList);
      }
    },
    closeDropdown () {
      if (this.isOpen) {
        this.isOpen = false;
        this.$refs.catalogDropdownList.scrollTop = 0;
        PerfectScrollbar.update(this.$refs.catalogDropdownList);
      }
    },
    disableDropdown () {
      this.isDisabled = true;
    },
    enableDropdown () {
      this.isDisabled = false;
    },
    onInputFocus () {
      this.isInputFocused = true;
    },
    onKeydownDown () {
      if (this.activeOptionIndex < this.options.length) {
        this.activeOptionIndex++;
      }
    },
    onKeydownUp () {
      if (this.activeOptionIndex > 0) {
        this.activeOptionIndex--;
      }
    },
    onKeyEnter () {
      const optionSelected = this.$el.querySelector('.catalogDropdown__option--active');
      if (optionSelected) {
        optionSelected.click();
      }
    },
    updateActiveOption (index) {
      this.activeOptionIndex = index;
    },
    resetActiveOption () {
      this.activeOptionIndex = -1;
    },
    selectOption (option) {
      this.setInput(option);
      this.$emit('selected', this.selected);
    },
    setInput (option) {
      this.enableDropdown();
      this.selected = option;
      this.searchFilter = this.selected;
      this.closeDropdown();
    },
    clearInput () {
      this.searchFilter = '';
      this.selected = {};
      this.closeDropdown();
    },
    reset () {
      this.clearInput();
      this.error = '';
      this.$emit('reset');
    },
    setError (error) {
      this.error = error;
    }
  },
  watch: {
    disabled () {
      this.isDisabled = this.disabled;
    },
    open () {
      this.isOpen = this.open;
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.catalogDropdown {
  --default-height-items: 8;
  --medium-height-items: 5;
  --small-height-items: 3;
  --extra-space: 2px;
  --catalog-dropdown-item-height: calc(48px + var(--extra-space));

  display: flex;
  z-index: $z-index__dropdown;
  flex-direction: column;
  width: 100%;

  &__container {
    position: relative;
    z-index: 2;
    width: calc(100% - 20px);
    margin-top: 16px;
  }

  &__input {
    width: 100%;
    height: 56px;
    padding: 16px 24px;
    border: 1px solid $neutral--300;
    border-radius: 4px;

    &.has-error {
      border: 1px solid $warning__border-color;
    }

    &--placeholder {
      color: $neutral--500;
    }
  }

  &__error {
    position: relative;
    margin-top: 4px;
    padding-left: 24px;
    font-size: 12px;
    line-height: 16px;

    &::before {
      content: '';
      display: block;
      position: absolute;
      top: -2px;
      left: 0;
      width: 18px;
      height: 18px;
      background-image: url(../../assets/icons/common/warning-icon.svg);
      background-repeat: no-repeat;
      background-position: center;
      background-size: 18px;
    }
  }

  &__close {
    display: block;
    position: absolute;
    top: 18px;
    right: 24px;
  }

  &.is-disabled {
    opacity: 0.24;

    .catalogDropdown__close {
      display: none;
    }
  }

  &.is-open {
    .catalogDropdown__input {
      border-bottom-right-radius: 0;
      border-bottom-left-radius: 0;
      border-top-color: transparent;
      border-right-color: transparent;
      border-left-color: transparent;
    }

    .catalogDropdown__list {
      border: transparent;
    }

    .catalogDropdown__container {
      border-radius: 4px;
      box-shadow: 0 0 0 2px $primary-color;
    }

    .catalogDropdown__extra {
      visibility: visible;
      opacity: 1;
      pointer-events: auto;
    }
  }

  &:not(.is-open) {
    .catalogDropdown__close {
      display: none;
    }

    .catalogDropdown__list,
    .catalogDropdown__extra {
      display: none;
    }

    &:hover {
      &:not(.is-disabled) {
        cursor: pointer;

        .catalogDropdown__close {
          display: block;
        }

        .catalogDropdown__input:hover {
          border-color: $text__color;
        }
      }
    }
  }
}

.catalogDropdown__list {
  visibility: hidden;
  position: relative;
  border: 1px solid $softblue;
  opacity: 0;
  background-color: $white;
  pointer-events: none;

  &.is-open {
    visibility: visible;
    opacity: 1;
    pointer-events: auto;
  }
}

.catalogDropdown__option {
  display: block;
  position: relative;
  width: 100%;
  padding: 12px 16px 12px 24px;
  overflow: hidden;
  color: $primary-color;
  text-decoration: none;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;

  &:not(:last-child) {
    border-bottom: 1px solid $softblue;
  }

  &--active {
    background-color: rgba($primary-color, 0.05);
    text-decoration: underline;
  }
}

.catalogDropdown__extra {
  visibility: hidden;
  padding: 16px 24px;
  border-top: 1px solid $softblue;
  border-bottom-right-radius: 4px;
  border-bottom-left-radius: 4px;
  opacity: 0;
  background-color: $white;
  pointer-events: none;

  &--text {
    color: $neutral--600;
  }
}

@media (max-height: 680px) {
  .catalogDropdown__list {
    &.is-height-limited {
      max-height: calc(var(--catalog-dropdown-item-height) * var(--small-height-items));
    }
  }
}

@media (min-height: 681px) and (max-height: 850px) {
  .catalogDropdown__list {
    &.is-height-limited {
      max-height: calc(var(--catalog-dropdown-item-height) * var(--medium-height-items));
    }
  }
}

@media (min-height: 851px) {
  .catalogDropdown__list {
    &.is-height-limited {
      max-height: calc(var(--catalog-dropdown-item-height) * var(--default-height-items));
    }
  }
}
</style>
