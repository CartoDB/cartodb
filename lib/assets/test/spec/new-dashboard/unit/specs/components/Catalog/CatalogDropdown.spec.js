import { shallowMount } from '@vue/test-utils';
import CatalogDropdown from 'new-dashboard/components/Catalog/CatalogDropdown';
import PerfectScrollbar from 'perfect-scrollbar';

const $t = key => key;

const defaultProperties = Object.freeze({
  title: 'Title',
  placeholderActive: 'CatalogDropdown.placeholder',
  placeholderInactive: 'CatalogDropdown.placeholder Disabled',
  options: [],
  open: true,
  disabled: false,
  limitHeight: false
});

describe('CatalogDropdown.vue', () => {
  function createCatalogDropdown (overridenComponentOptions) {
    return shallowMount(CatalogDropdown, {
      mocks: {
        $t
      },
      propsData: {
        ...defaultProperties
      },
      slots: {
        extra: 'Extra Content'
      },
      ...overridenComponentOptions
    });
  }

  beforeAll(() => {
    spyOn(PerfectScrollbar, 'initialize').and.returnValue('initialized');
    spyOn(PerfectScrollbar, 'update').and.returnValue('updated');
    spyOn(PerfectScrollbar, 'destroy').and.returnValue('destroyed');
  });

  it('should render correct contents', () => {
    const catalogDropdown = createCatalogDropdown();
    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should render inactive placeholder if dropdown is inactive', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        ...defaultProperties,
        disabled: true,
        placeholderInactive: 'Placeholder Disabled'
      }
    });
    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be opened by default if open is true', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        ...defaultProperties,
        open: true
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be closed by default if open is false', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        open: false
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be disabled by default if disabled is true', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        disabled: true
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  it('should be enabled by default if disabled is false', () => {
    const catalogDropdown = createCatalogDropdown({
      propsData: {
        disabled: false
      }
    });

    expect(catalogDropdown).toMatchSnapshot();
  });

  describe('Items', () => {
    it('should show up to 8 items if the screen height by default', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9'];

      window.innerHeight = 1200;

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown.vm.maxItemsScroll).toEqual(8);
    });

    it('should show up to 3 items if the screen height is "small"', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9'];

      window.innerHeight = 300;

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown.vm.maxItemsScroll).toEqual(3);
    });

    it('should show up to 5 items if the screen height is "medium"', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9'];

      window.innerHeight = 700;

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown.vm.maxItemsScroll).toEqual(5);
    });
  });

  describe('Search', () => {
    it('should be visible when there is an items overflow', () => {
      const options = ['o_1', 'o_2', 'o_3', 'o_4', 'o_5', 'o_6', 'o_7', 'o_8', 'o_9', 'o_10'];

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          ...defaultProperties,
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown).toMatchSnapshot();
    });

    it('should be hidden if there is no items overflow', () => {
      const options = ['o_1', 'o_2', 'o_3'];

      const catalogDropdown = createCatalogDropdown({
        propsData: {
          ...defaultProperties,
          options,
          open: true,
          disabled: false,
          limitHeight: false
        }
      });

      expect(catalogDropdown).toMatchSnapshot();
    });
  });
});
