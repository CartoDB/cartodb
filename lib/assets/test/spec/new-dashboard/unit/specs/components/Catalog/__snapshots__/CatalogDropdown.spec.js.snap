// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogDropdown.vue Search should be hidden if there is no items overflow 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label">Title</span>
  <div class="catalogDropdown__container">
    <div><input type="text" placeholder="CatalogDropdown.placeholder" class="text is-caption catalogDropdown__input"></div>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container">
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_1"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_2"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_3"></catalogdropdownitem-stub>
      </li>
    </ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue Search should be visible when there is an items overflow 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label">Title</span>
  <div class="catalogDropdown__container">
    <div><input type="text" placeholder="CatalogDropdown.placeholder" class="text is-caption catalogDropdown__input"></div>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container ps-active-y">
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_1"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_2"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_3"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_4"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_5"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_6"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_7"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_8"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_9"></catalogdropdownitem-stub>
      </li>
      <li class="catalogDropdown__option text is-caption">
        <catalogdropdownitem-stub option="o_10"></catalogdropdownitem-stub>
      </li>
    </ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be closed by default if open is false 1`] = `
<div class="catalogDropdown"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input"><span class="catalogDropdown__input--placeholder"></span>
      <!---->
    </div>
    <!---->
    <ul class="catalogDropdown__list"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be disabled by default if disabled is true 1`] = `
<div class="catalogDropdown is-disabled"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input"><span class="catalogDropdown__input--placeholder"></span>
      <!---->
    </div>
    <!---->
    <ul class="catalogDropdown__list"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be enabled by default if disabled is false 1`] = `
<div class="catalogDropdown"><span class="title is-caption catalogDropdown__label"></span>
  <div class="catalogDropdown__container">
    <div class="text is-caption catalogDropdown__input"><span class="catalogDropdown__input--placeholder"></span>
      <!---->
    </div>
    <!---->
    <ul class="catalogDropdown__list"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should be opened by default if open is true 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label">Title</span>
  <div class="catalogDropdown__container">
    <div><input type="text" placeholder="CatalogDropdown.placeholder" class="text is-caption catalogDropdown__input"></div>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should render correct contents 1`] = `
<div class="catalogDropdown is-open"><span class="title is-caption catalogDropdown__label">Title</span>
  <div class="catalogDropdown__container">
    <div><input type="text" placeholder="CatalogDropdown.placeholder" class="text is-caption catalogDropdown__input"></div>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;

exports[`CatalogDropdown.vue should render inactive placeholder if dropdown is inactive 1`] = `
<div class="catalogDropdown is-disabled is-open"><span class="title is-caption catalogDropdown__label">Title</span>
  <div class="catalogDropdown__container">
    <div><input type="text" placeholder="Placeholder Disabled" disabled="disabled" class="text is-caption catalogDropdown__input"></div>
    <!---->
    <ul class="catalogDropdown__list is-open ps-container"></ul>
    <div class="catalogDropdown__extra">Extra Content</div>
  </div>
</div>
`;
